src/sql/workbench/contrib/notebook/browser/notebook.contribution.ts

```diff
@@ -658,3 +663,74 @@ configurationRegistry.registerConfiguration({
+
+export class NotebookEditorOverrideContribution extends Disposable implements IWorkbenchContribution {
+
+	private _registeredOverrides = new DisposableStore();
+
+	constructor(
+		@ILogService private _logService: ILogService,
+		@IEditorService private _editorService: IEditorService,
+		@IEditorOverrideService private _editorOverrideService: IEditorOverrideService,
+		@IModeService private _modeService: IModeService
+	) {
+		super();
+		this.registerEditorOverride();
+	}
+
+	private registerEditorOverride(): void {
+		// Refresh the editor overrides whenever the languages change so we ensure we always have
+		// the latest up to date list of extensions for each language
+		this._modeService.onLanguagesMaybeChanged(() => {
+			this._registeredOverrides.clear();
+			// List of language IDs to associate the query editor for. These are case sensitive.
+			NotebookEditorInputAssociation.languages.map(lang => {
+				const langExtensions = this._modeService.getExtensions(lang);
+				if (langExtensions.length === 0) {
+					return;
+				}
+				// Create the selector from the list of all the language extensions we want to associate with the
+				// query editor (filtering out any languages which didn't have any extensions registered yet)
+				const selector = `*{${langExtensions.join(',')}}`;
+				this._registeredOverrides.add(this._editorOverrideService.registerContributionPoint(
+					selector,
+					{
+						id: NotebookEditor.ID,
+						label: NotebookEditor.LABEL,
+						describes: (currentEditor) => currentEditor instanceof FileNotebookInput,
+						priority: ContributedEditorPriority.builtin
+					},
+					{},
+					(resource, options, group) => {
+						const fileInput = this._editorService.createEditorInput({
+							resource: resource
+						}) as FileEditorInput;
+						// Try to convert the input, falling back to just a plain file input if we're unable to
+						const newInput = this.tryConvertInput(fileInput, lang) ?? fileInput;
+						return { editor: newInput, options: options, group: group };
+					},
+					(diffEditorInput, options, group) => {
+						// Try to convert the input, falling back to the original input if we're unable to
+						const newInput = this.tryConvertInput(diffEditorInput, lang) ?? diffEditorInput;
+						return { editor: newInput, options: options, group: group };
+					}
+				));
+			});
+		});
+	}
+
+	private tryConvertInput(input: IEditorInput, lang: string): IEditorInput | undefined {
+		const langAssociation = languageAssociationRegistry.getAssociationForLanguage(lang);
+		const notebookEditorInput = langAssociation?.syncConvertinput?.(input);
+		if (!notebookEditorInput) {
+			this._logService.warn('Unable to create input for overriding editor ', input instanceof DiffEditorInput ? `${input.primary.resource.toString()} <-> ${input.secondary.resource.toString()}` : input.resource.toString());
```

Comment:
What does `<->` do?

Reply:
That's just a text string (diff editors don't have a resource, just primary and secondary, so I wanted to just display it nicely)